proposals/LP0004-project-governance.md

```diff
@@ -0,0 +1,393 @@
+
+The LLVM Foundation will maintain a voter database that identifies voters by name, email, and the accounts used on services used by the LLVM project. The voter database will only be used for election-related purposes. A full privacy policy that complies with appropriate laws will be published, and it will include a process for removing your personal information from the voter database. To be a _voting contributor_ an individual must register with the LLVM Foundation voter database, and meet eligibility requirements.
+
+An individual may be deemed ineligible to vote for violations of the community Code of Conduct. Acting fraudulently or any other method of subverting an LLVM election is a violation of the Code of Conduct.
+
+Voter eligibility is based on recent community participation in the form of _meaningful contributions_. A _meaningful contribution_ may be a pull request contribution, or it could be other activities that help the community function like screening issues, moderating community spaces, hosting community events, or participating in code reviews.
+
+7 days before voting begins a list of eligible voters is constructed from the registered voters including only those that have participated in LLVM community interactions in the preceding 12 months.
+
+Using the methodologies defined in Appendix 2, and only looking at the main LLVM repository, we have 1667 eligible _voting contributors_ based on code contributions alone. This number should expand as we incorporate participation in code reviews, issues, and discussion forums.
+
+
+#### Alternative Considered: Commit Access
+
+Alternatively, the _voting contributor_ base could be defined by commit access. This methodology has two flaws. One is that our commit access list gets stale, and is unwieldy to maintain. The other problem is that with the move to GitHub PRs, commit access is significantly less important. We may begin getting more contributions from users that never bother to request commit privileges.
+
+
+### Area Teams
+
+The second structure this proposal establishes is a collection of _area teams_. Members of _area teams_ are not required or expected to be experts in the area the team is responsible for. There are no skill or experience requirements to be on an _area team_.
+
+_Area teams_ have three main responsibilities.
+
+First, they are responsible for electing from among themselves a team secretary who will take notes of any team meetings and a team chair who represents the team on the _project council_.
+
+Second, _area teams_ are responsible for maintaining an up-to-date and comprehensive list of code owners for their area of the project. They can nominate any individual they deem appropriate as code owner of any area they are responsible for. The role of _code owner_ remains a volunteer role, and any individual can accept, decline, or resign the role for themselves as they feel appropriate.
```

Comment:
Playing devil's advocate: Who says code owners are required to be technical experts?

The developer policy doesn't state that, and based on the role description in the developer policy it doesn't even seem like that requirement should be expected.

This document doesn't intend to change code owners roles or process other than having someone responsible for curating the lists. Just as today when someone "nominates" a code owner by posting on Discourse, I see the area teams as having the role of posting out that they want to nominate code owners and seeing if the community objects. No voting required.

Reply:
> based on the role description in the developer policy it doesn't even seem like that requirement should be expected.

I strongly disagree with that statement. It may not be in the writing, but back when we were writing the document, I believe no one thought a code owner would not be a technical expert on the area they are code owners. I may be wrong, and would love to be proven so, but until we actualy find that this was the case, I would not assume so.

This document makes that assumption and encodes into writing, which is not correct. We need to make sure this is the actual consensus before encoding it in writing.

Reply:
I think you just made my point for me. The developer policy _does not_ place any requirement on the role, but _in practice_ people tend to be at least technically strong in the area.

I can think of at least one instance that I was directly impacted by where a self-appointed code owner was not particularly familiar with the area they owned, and they didn't do a particularly great job of being a code owner either. At one point I even confronted them about it and got an unsatisfactory response. The problem as I see it wasn't that they were not an expert, but was rather that they didn't do anything to help get patches reviewed.

>This document makes that assumption and encodes into writing, which is not correct. We need to make sure this is the actual consensus before encoding it in writing.

The document only encodes the statement as it applies to the _area teams_. I don't make any change to the policy for code owners. Code owners can be experts (or not) on a case-by-case basis subject to the same process (or lack thereof) that we've always had. If the Developer Policy needs to be revised to codify requirements or an election process for code owners I see that as a separate problem.

Reply:
> The document only encodes the statement as it applies to the _area teams_. I don't make any change to the policy for code owners.

Yes, however, the document states: _"They can nominate any individual they deem appropriate as code owner of any area they are responsible for."_

What does _"nominate"_ means here?

Today, anyone can nominate anyone to be a code owner, usually themselves, and if there's consensus, they become the code owner. But to become a code owner, they need consensus.

If area teams have the same "rights", then the text is redundant and unnecessary. Being there, could be interpreted as "more rights" and is confusing at best.

If they actually have "more rights", for instance, they can nominate and persist the change without consensus, then this is basically the same as saying they can add/change/remove code owners at will. That is not the right level of power I'd be comfortable with for an area team.

My understanding is that area teams have administrative oversight, not ultimate power. They work within the framework to resolve issues, not around it, not behind it. The way it's written gives them unfettered power to change code owners at whim.

Please understand, I'm not assuming malice here. But bias and good intentions are the source wider structural changes than malice, because it goes undetected and the direct effects are usually interpreted as "side effects".

Every time we encode something in a document, we're opening new possibilities that go far beyond what we intended, so we need to be extra careful on what and how we write. 

Like a compiler pass, we should only allow what has been proven right, not what we think "might work for some cases".